### PR TITLE
Build: A basic monorepo setup

### PR DESCRIPTION
**Pull request by blowery**
Everybody get on board!

This PR adds a lerna config to turn Calypso into a monorepo, where we can host the source for multiple npm modules. Initially, this will be the Muriel UI components. We're likely going to bring in i18n-calypso in the near future. 

Work so far:
* add basic package config
* add lerna, update shinkwrap

To Do:
- [ ] How do we want to handle building / watching / rebuilding?
- [ ] How to support tests w/ jest
- [ ] Write up the expected dev flow with this setup

### CODE REVIEW

**Comment by blowery** [comment]
Fun things so far:
* Once you start using `lerna` you have to stop using `npm` directly. All of our `npm ci` calls need to turn into `lerna bootstrap` (I think).

**Comment by belcherj** [comment]
Nice!!!  This is something I worked on during my trial.  Memory lane!

**Comment by belcherj** [comment]
https://github.com/Automattic/wp-calypso/pull/3300

File: client/reader/following/main.jsx [hunk 1]
```diff
@@ -53,15 +54,17 @@ const FollowingStream = props => {
...
-			<div className="search-stream__blank-suggestions">
-				{ suggestionList &&
-					props.translate( 'Suggestions: {{suggestions /}}.', {
-						components: {
-							suggestions: suggestionList,
-						},
-					} ) }
-				&nbsp;
-			</div>
+			<Muriel>
```
Comment (blowery) [inline client/reader/following/main.jsx]: this is just a test to prove it works

File: Dockerfile [hunk 2]
```diff
@@ -44,6 +35,7 @@ RUN        npm ci --only=production
 # not trigger additional install as part of the build
 # in the following step.
 COPY       . /calypso/
+RUN        npm ci --only-production
```
Comment (sirreal) [inline Dockerfile]: Typo here: `--only=production`.

The docs on `npm ci` are pretty bad, it's unclear to me whether it actually respects this command which we've kept from `npm install`.

File: bin/build-packages.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+glob( 'packages/*/package.json', ( err, matches ) => {
+	if ( err ) { console.error( err ); return; }
+
+	matches.forEach( match => {
+		const dir = path.dirname( match );
+		const inputDir = path.join( dir, 'src' );
+		const outputDir = path.join( dir, 'dist', 'esm' );
+		console.log( `Building ${ dir }...` );
+		execSync( `node ./node_modules/.bin/babel -d ${outputDir} ${inputDir}`, {
```
Comment (blowery) [inline bin/build-packages.js]: tried this with npx but it seemed to try to download babel@5?

File: lerna.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "packages": [
```
Comment (blowery) [inline lerna.json]: should we move calypso into `apps/calypso` and add `apps/*` here? eventually?

File: packages/muriel-base-components/package.json [hunk 5]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "muriel-base-components",
```
Comment (ockham) [inline packages/muriel-base-components/package.json]: A bit academic as long as we don't publish, but to set good precedent: maybe scope, i.e. `@automattic/muriel-base-components`?

File: bin/build-packages.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+			env: Object.assign( {}, process.env, { CALYPSO_CLIENT: 'true' } )
```
Comment (ockham) [inline bin/build-packages.js]: This means transpile to ES5/ESM (modules), right? We should really rename that flag :slightly_frowning_face:

File: packages/muriel-base-components/package.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  "version": "1.0.0",
+  "description": "The Muriel Base Components",
+  "main": "dist/cjs/index.js",
```
Comment (ockham) [inline packages/muriel-base-components/package.json]: No CJS build yet -- should we remove for now?

File: bin/build-packages.js (hunk 6 above)
Comment (blowery) [reply bin/build-packages.js]: yeah, it means a bunch of stuff right now.

Comment (ockham) [reply bin/build-packages.js]: Maybe add an inline comment for the time being to explain it?

File: Dockerfile [hunk 8]
```diff
@@ -44,6 +35,7 @@ RUN        npm ci --only=production
 # not trigger additional install as part of the build
 # in the following step.
 COPY       . /calypso/
+RUN        npm ci --only=production
 RUN        touch node_modules
```
Comment (ockham) [inline Dockerfile]: Do we still need this line? (If we don't, we should also update the above comment ^)

File: bin/build-packages.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+		} );
+	 } );
+	 console.log('Done building packages');
+} );
```
Comment (ockham) [inline bin/build-packages.js]: And alternative approach to this file would be adding `build` `scripts` to individual files' `package.json`s (which arguably keeps 'em a bit more self-contained), and for the app-level build step, [`lerna run build`](https://lernajs.io/#command-run).

**Review by ockham — APPROVED** [review]
Left a couple of notes, but by and large, I'd say -- let's merge this soon? Looks like a solid basis for iteration, and I'd be eager to use this for #27123.

File: packages/muriel-base-components/package.json (hunk 7 above)
Comment (blowery) [reply packages/muriel-base-components/package.json]: yeah, i guess. this is mostly exploratory so far.

File: bin/build-packages.js (hunk 9 above)
Comment (blowery) [reply bin/build-packages.js]: the thing i ran into early was that `npx babel ..` wouldn't find the babel sitting in the node_modules a directory up. We could just go after it directly (`node ../node_modules/.bin/babel ...`)...

Comment (ockham) [reply bin/build-packages.js]: Could be resolved by [hoisting](https://github.com/lerna/lerna/blob/master/doc/hoist.md)?

> Binaries from those common packages are symlinked to individual package `node_modules/.bin` directories, so that `package.json` scripts continue to work unmodified.

Comment (blowery) [reply bin/build-packages.js]: possibly, but it means a new thing to use and learn, plus hoisting seems to be somewhat experimental.

File: Dockerfile [hunk 10]
```diff
@@ -20,31 +20,12 @@ ENV        NODE_PATH=/calypso/server:/calypso/client
...
-# However, time is being spent in the build step on
-# `install-if-deps-outdated`. This is because in the
-# following COPY, the npm-shrinkwrap mtime is being
-# updated, which is confusing `install-if-deps-outdated`.
-# Touch after copy to ensure that this layer will
-# not trigger additional install as part of the build
-# in the following step.
 COPY       . /calypso/
-RUN        touch node_modules
+RUN        npm ci --only=production
```
Comment (blowery) [inline Dockerfile]: @sirreal can you see any way around this? with `file:` refs in the package.json, I don't think we can multi-layer this like we did.

**Comment by blowery** [comment]
I'm going to land this as is. We can revisit after it lands.

File: packages/muriel-base-components/src/button/index.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @export
+ * @param {*} props The props to use
+ * @returns {React.ReactElement} a rendered Button
+ */
+export class Button extends Component {
+	render() {
+		return <button className="muriel-button">{ this.props.children }</button>;
+	}
+}
```
Comment (dmsnell) [inline packages/muriel-base-components/src/button/index.js]: why is this here and not in `components/button`?

**Comment by sirreal** [comment]
> How to support tests w/ jest

I've been meaning to experiment with this but haven't had time yet. Something that looks really promising is `projects`:

> When the `projects` configuration is provided with an array of paths or glob patterns, Jest will run tests in all of the specified projects at the same time. This is great for monorepos or when working on multiple projects at the same time.
> 
> ```json
> {
>   "projects": ["<rootDir>", "<rootDir>/examples/*"]
> }
> ```

https://jestjs.io/docs/en/configuration.html#projects-array-string-projectconfig

File: Dockerfile (hunk 10 above)
Comment (dmsnell) [reply Dockerfile]: with `file:` refs will it matter? those aren't "installed" right - ?

in other words, why does a cached `npm install` which fetches the external libs invalidate a build that also resolves certain modules locally?
